File: theta.jl
using ArgParse
using CSV
using DataFrames

# include("sinkhorn_mul.jl")
# include("sinkhorn_mul_logstab.jl")
include("sinkhorn_explmul_logstab.jl")

const re = r"([\+\-]?\d{1,}\.?\d{0,}[eE]?[\+\-]?\d{0,})"

function parse_commandline()
    s = ArgParseSettings()
    @add_arg_table s begin
        "--resultsfile"
            help="filename of the result"
        "--exfile"
            help="filename of the excitations"
        "--orbitals"
            help="name of the orbital directory"
        "--dir"
            help="name of general directory"
    end
    return parse_args(s)
end

function cube_to_array(filename,vals,dV,retrn)
    natoms = 0
    xlim = 0.0
    ylim = 0.0
    zlim = 0.0
    nx = 0
    ny = 0
    nz = 0
    open(filename, "r") do datfile
        for (i,line) in enumerate(eachline(datfile))
    		if i < 3
    			continue
    		elseif i==3
    			tmp = eachmatch(re,line)
                tmp = [ match.match for match in eachmatch(re, line)]
    			natoms = parse(Int64,tmp[1])
    			xlim, ylim, zlim = parse(Float64, tmp[2]), parse(Float64, tmp[3]), parse(Float64, tmp[4])
    		elseif 4 <= i && i <=6
                tmp = [ parse(Float64,match.match) for match in eachmatch(re, line)]
    			# n = parse(Int64,tmp[1])
                n = floor(Int64,tmp[1])
    			if tmp[2] != 0
    				nx = n
    				dx = tmp[2]
                    dV[1] = tmp[2]
                end
    			if tmp[3] != 0
    				ny = n
    				dy = tmp[3]
                    dV[2] = tmp[3]
                end
    			if tmp[4] != 0
    				nz = n
    				dz = tmp[4]
                    dV[3] = tmp[4]
                end
    		elseif 6 < i && i <= 6+abs(natoms)
    			continue
    		elseif i==6+abs(natoms)+1 && natoms < 0
    			continue
    		else
                append!(vals,[ parse(Float64,match.match) for match in eachmatch(re, line)])
            end
        end
    end

    grid = zeros((nx*ny*nz,3))
	for i in 1:nx
		for j in 1:ny
			for k in 1:nz
                n = k + (j-1)*nz + (i-1)*nz*ny
                grid[n,1] = xlim + (i-1)*dV[1]
                grid[n,2] = ylim + (j-1)*dV[2]
                grid[n,3] = zlim + (k-1)*dV[3]
                # append!(grid,Array{Float64}[[xlim + (i-1)*dx, ylim + (j-1)*dy, zlim + (k-1)*dz]])
            end
        end
    end

    if (retrn)
        return grid
    end
end

function wasserstein(filenamea,filenameb,eps=.5)
    # initialise empty arrays
    vals1 = zeros(0)
    vals2 = zeros(0)
    dV = zeros(3)

    println("stats for reading of orbitals")
    @time grid=cube_to_array(filenamea,vals1,dV,true)
    @time cube_to_array(filenameb,vals2,dV,false)
    println(size(grid))
    # now that all integrals are >= 0.99, the error introduced here is small but it's needed for the Sinkhorn
    vals1 .= vals1./sqrt(sum(vals1.^2))#*prod(dV))
    vals2 .= vals2./sqrt(sum(vals2.^2))#*prod(dV))
    # println("check valsums are equal: ",sum(vals1.^2)," and ",sum(vals2.^2))
    println("stats for Sinkhorn")
    cost,res =  sinkhorn_explmul_logstab(grid,grid,vals1.^2 .+ 1e-16,vals2.^2 .+ 1e-16,eps)
    # return 1.0
    return cost
end

function main()
    print("WARNING: THIS VERSION ONLY CALCULATES THE COST PORTION AND DOES NOT CORRECT FOR SELF-CORRELATION.")

    isturbomole = true

    args = parse_commandline()
    resultsfile = args["resultsfile"]
    exfile = args["exfile"]
    orbitals = args["orbitals"]
    dir = args["dir"]

    open(resultsfile, "w") do resfile
        write(resfile,"Molecule,Functional,Excitation,Theta1,Theta2\n")
        exdata = CSV.read(exfile, DataFrame)
        for i in 1:nrow(exdata)
            failed = false
            molecule = exdata[i,"Molecule"]
            functional = exdata[i,"Functional"]
            excitation = exdata[i,"Excitation"]
            ncntrib = exdata[i,"no contr"]
            if (ncntrib == 0)
                continue
            end

            versions = 1
            if occursin("Delta",excitation)
                versions = 2
            end
            Theta = zeros(versions)

            for j in 1:versions
                for k in 1:ncntrib
                    phia = exdata[i,"occ$k"]
                    phib = exdata[i,"virt$k"]
                    c = exdata[i,"contr$k"]
                    if (isturbomole)
                        phia = replace(phia, " " => "")
                        phib = replace(phib, " " => "")

                        if occursin("e",phia) && occursin("e",phib)
                            if j==1
                                phia = phia * "1"
                                phib = phib * "1"
                            end
                            if j==2
                                phia = phia * "1"
                                phib = phib * "2"
                            end
                        elseif occursin("e",phia)
                            if j==1
                                phia = phia * "1"
                            end
                            if j==2
                                phia = phia * "2"
                            end
                        elseif occursin("e",phib)
                            if j==1
                                phib = phib * "1"
                            end
                            if j==2
                                phib = phib * "2"
                            end
                        end

                        filenamea = dir * molecule * "/" * functional * "/" * orbitals * "/" * phia * ".cub"
                        filenameb = dir * molecule * "/" * functional * "/" * orbitals * "/" * phib * ".cub"
                    else
                        filenamea = dir * molecule * "/" * molecule * "_" * functional * ".mo" * phia * ".cube"
                        filenameb = dir * molecule * "/" * molecule * "_" * functional * ".mo" * phib * ".cube"
                    end
                    println("Molecule: ",molecule,", Excitation: ", excitation,", Functional: ", functional,", Phia: ", phia,", Phib: ", phib)
                    if (isfile(filenamea) && isfile(filenameb))
                        th = wasserstein(filenamea,filenameb)
                        println("Theta: ",th," with contribution: ",c)
                        Theta[j] = Theta[j] + c*th
                    else
                        println("failed: ",molecule,", ",excitation,", ",functional,", ",phia,", ",phib)
                        failed = true
                    end

                end # ncntrib
            end # versions
            if (!failed)
                 # write row
                 for i=1:versions
                     if i==2
                         write(resfile,molecule*functional*excitation*",,$(Theta[i])\n")
                     else
                         write(resfile,molecule*functional*excitation*",$(Theta[i])\n")
                     end
                 end
            end
        end
    end
end

main()

File: theta.py
import torch
from geomloss import SamplesLoss
import numpy as np
import numpy.linalg
import utils
import argparse
import csv
import os


parser = argparse.ArgumentParser(description="Computation of Theta descriptor")
parser.add_argument('--dir', default=os.getcwd() + '/', help="working directory for orbitals")
parser.add_argument('--csvfile', help='name of CSV file with excitation data')
parser.add_argument('--orbitals', help='home directory of orbitals')
parser.add_argument('--results', help='results directory')
parser.add_argument('--format', help='which program is used to obtain orbitals')


def sinkhorndiv(phia,phib):
	''' Calculates the Sinkhorn divergence of |phi_a|**2 to |phi_b|**2 based on the
	geomloss package.
	'''
	grid,vals1,stats1 = utils.cube_to_array(phia)
	_,vals2,stats2 = utils.cube_to_array(phib)
	vals1 = vals1/np.sqrt(np.sum(vals1**2))
	vals2 = vals2/np.sqrt(np.sum(vals2**2))
	vals1 = vals1**2+1e-16
	vals2 = vals2**2+1e-16
	mxdist = np.linalg.norm(grid[0,:]-grid[-1,:])**2
	eps = mxdist/1e4
	Loss = SamplesLoss("sinkhorn", p=2, blur=eps, scaling=0.8)
	
	v1 = torch.from_numpy(vals1)
	v2 = torch.from_numpy(vals2)
	x = torch.from_numpy(grid)
	y = torch.clone(x)
	return Loss(v1,x,v2,y).item(),eps


if __name__=="__main__":
	use_cuda = torch.cuda.is_available()
	dtype = torch.cuda.FloatTensor if use_cuda else torch.FloatTensor
	
	args = parser.parse_args()

	# initialise where the results are to be stored
	resultsfile = args.results

	with open(resultsfile, 'w') as wfile:
		writer = csv.DictWriter(wfile, fieldnames=["Molecule", "Excitation", "Functional", "Theta1", "Theta2", "eps"])
		writer.writeheader()
		with open(args.csvfile,'r') as file:
			reader = csv.DictReader(file)
			# every line corresponds to one excitation
			for line in reader:
				failed=False
				if line["no contr"] == '':
					continue
				molecule = line["\ufeffMolecule"]
				# molecule = line["Molecule"]
				excitation = line["Excitation"]
				functional = line["Functional"]
				if functional == "CAM-B3LYP":
					functional = "CAMB3LYP"
				no_contr = line["no contr"]
				print("Moleule:",molecule)
				print("Excitation:",excitation)
				print("Functional:",functional)
				print("Contributions:")

				versions = 1
				if "Delta" in excitation or "Sigma-" in excitation or "Sigmau-" in excitation:
					versions = 2
				Theta = np.zeros(versions)

				for i in range(versions):
					for j in range(int(no_contr)):
						phia = line["occ" + str(j + 1)]
						phib = line["virt" + str(j + 1)]
						c = line["contr" + str(j + 1)]
						if args.format=='turbomole':
							if "e" in phia and "e" in phib:
								if i==0:
									phia += "1"
									phib += "1"
								if i==1:
									phia += "1"
									phib += "2"
								if i==2:
									phia += "2"
									phib += "1"
								if i==3:
									phia += "2"
									phib += "2"
							elif "e" in phia:
								if i==0:
									phia += "1"
								if i==1:
									phia += "2"
							elif "e" in phib:
								if i==0:
									phib += "1"
								if i==1:
									phib += "2"


						if args.format=='orca':
							filenamea = args.dir + molecule + "/" + args.orbitals + molecule + "_" + functional + ".mo" + phia + ".cube"
							filenameb = args.dir + molecule + "/" + args.orbitals + molecule + "_" + functional + ".mo" + phib + ".cube"
						elif args.format=='turbomole':
							phia = phia.replace(" ","")
							phib = phib.replace(" ","")
							filenamea = args.dir + molecule + "/" + functional + args.orbitals + phia + ".cub"
							filenameb = args.dir + molecule + "/" + functional + args.orbitals + phib + ".cub"

						if os.path.isfile(filenamea) and os.path.isfile(filenameb):
							th,eps = sinkhorndiv(filenamea, filenameb)
							print(f"Theta %6.4f with contribution %5.3f"%(th,float(c)))
							Theta[i] += th*float(c)
						else:
							print("Failed:",molecule, excitation, functional, phia, phib, versions)
							print("Looked for files:")
							print(filenamea)
							print(filenameb)
							failed = True
							eps = 1000
							break


				if not failed:
					if (versions==1):
						writer.writerow({"Molecule": molecule, "Excitation": excitation, "Functional": functional, "Theta1": Theta[0], "eps":eps})
					else:
						writer.writerow({"Molecule": molecule, "Excitation": excitation, "Functional": functional, "Theta1": Theta[0], "Theta2": Theta[1], "eps":eps})
